Add example notebook with JUMP UMAP processing

d33bs opened the pull request:
<!-- _modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_
_referenced with modifications from [pycytominer](https://github.com/cytomining/pycytominer/blob/master/.github/PULL_REQUEST_TEMPLATE.md)_ -->

# Description

This PR adds an example notebook to help demonstrate coSMicQC use with JUMP by performing a UMAP analysis. The output generally show where outliers occur within the embeddings from a sample of BR00117012 sourced from the cpg0000-jump-pilot JUMP dataset.

Additional notes:
- I took a moment to update the `cosmicqc_in_a_nutshell` notebook which recently was reorganized / was lacking display of outlined cells within the CytoDataFrame.
- As a result of this modification I also noticed some unnecessary print statements which I removed.
- The coverage was changed I believe due to these print statements being removed.
- I've double checked locally and saw that the interactive Bokeh plots render properly through Sphinx (which means they should render within GitHub Pages).

Thanks for any feedback you may have!

Closes #23 
Closes #83
Closes #90 

## What kind of change(s) are included?

- [x] Documentation (changes docs or other related content)
- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read and followed the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] I have searched for existing content to ensure this is not a duplicate.
- [x] I have performed a self-review of these additions (including spelling, grammar, and related).
- [x] These changes pass all pre-commit checks.
- [x] I have added comments to my code to help provide understanding
- [ ] I have added a test which covers the code changes found within this PR
- [x] I have deleted all non-relevant text in this pull request template.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+
+# check if we already have normalized data
+if not pathlib.Path(parquet_sampled_with_outliers).is_file():
+    # set a fraction for sampling
```
might be helpful to add comment describing why 0.44

d33bs replied:
Thanks, will do!

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    # group by metadata_well for all features then sample
+    # the dataset by a fraction.
+    df_features = (
+        df_features.groupby(["Metadata_Well"])[df_features.columns]
```
```suggestion
        df_features.groupby(["Metadata_Well"])
```

I don't think this is necessary

d33bs replied:
Thanks for catching this - I agree it does look a bit unusual for syntax. I found I had to add the column selection to avoid a warning from Pandas regarding how groupby is used here. Going to retain this code for now. See here for more: https://stackoverflow.com/questions/77969964/deprecation-warning-with-groupby-apply

A warning in the output of the cell otherwise shows:

DeprecationWarning:
DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    df_feature_selected_with_cqc_outlier_data.to_parquet(parquet_sampled_with_outliers)
+
+else:
+    df_feature_selected_with_cqc_outlier_data = pd.read_parquet(
```
When you reran this to confirm the non-duck shape, did you start from scratch? (or did you use the existing parquet file)

d33bs replied:
Thanks, I did re-run everything in full to confirm. I added the checks here to help speed up subsequent steps and as a safety from kernel crashing.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    )
+
+    # join the sampled feature data with the cosmicqc outlier data
+    df_feature_selected_with_cqc_outlier_data = df_features.merge(
```
this is a bit confusing - I don't think these are feature selected profiles? Aren't they just the cytotable processed features (prior to pycytominer processing) merged with outlier info?

d33bs replied:
Thanks for catching this, yes that's accurate. I shifted the location of this code and forgot to change the name. I'll get this updated.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+
+# prepare data for normalization and feature selection
+# by removing cosmicqc and analaysis focused columns.
+df_for_normalize_and_feature_select = df_feature_selected_with_cqc_outlier_data[
+    [
+        col
+        for col in df_feature_selected_with_cqc_outlier_data.columns.tolist()
+        if "cqc." not in col and "analysis." not in col
+    ]
+]
```
```suggestion
df_for_normalize_and_feature_select = df_feature_selected_with_cqc_outlier_data.loc[:, df_features.columns]
```

Won't this work? (simpler)

d33bs replied:
Thanks, yes, simpler, adding it in!

d33bs replied:
Ended up referencing the CytoTable merged single cells parquet file schema for this, to conserve on data reads.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+if not pathlib.Path(parquet_pycytominer_normalized).is_file():
+    # normalize the data using pcytominer
+    df_pycytominer_normalized = pycytominer.normalize(
+        profiles=df_for_normalize_and_feature_select,
```
```suggestion
        profiles=df_for_normalize_and_feature_select,
        samples="Metadata_control_type == 'negcon'",
```

I think this is what we actually want? Samples normalized to the negative control cells? @jenna-tomkinson - please check me

d33bs replied:
Thanks - I might have missed this for processing with CytoTable. What table would this data come from?

jenna-tomkinson replied:
Correct! It is `samples="Metadata_control_type == 'negcon'"` that we use for normalizing.

d33bs replied:
Thanks both! I'll add in the pycytominer annotation step using https://open.quiltdata.com/b/cellpainting-gallery/tree/cpg0000-jump-pilot/source_4/workspace/metadata/platemaps/2020_11_04_CPJUMP1/platemap/JUMP-Target-1_compound_platemap.txt as a platemap for reference, then add in the `samples` parameter for normalization.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+len(df_for_normalize_and_feature_select.columns)
+
+# +
+parquet_pycytominer_normalized = "./BR00117012_normalized.parquet"
```
```suggestion
parquet_pycytominer_normalized = f"./{example_plate}_normalized.parquet"
```

I recommend doing this (and defining `example_plate` earlier)

d33bs replied:
Thanks, I'll add that in!

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    )
+
+# +
+parquet_pycytominer_feature_selected = "./BR00117012_feature_select.parquet"
```
```suggestion
parquet_pycytominer_feature_selected = f"./{example_plate}_feature_select.parquet"
```

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+
+
+# +
+def generate_umap_embeddings(
```
probably should define earlier in this notebook to satisfy PEP

d33bs replied:
Thanks, I'll move this to an earlier point in the notebook.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+parquet_pycytominer_feature_selected = "./BR00117012_feature_select.parquet"
+
+# check if we already have feature selected data
+if not pathlib.Path(parquet_pycytominer_feature_selected).is_file():
```
Just confirming again that we reran this fresh

d33bs replied:
Thanks for double checking, yes, this was re-run fresh.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+if not pathlib.Path(parquet_pycytominer_feature_selected).is_file():
+    # feature select normalized data using pycytominer
+    df_pycytominer_feature_selected = pycytominer.feature_select(
+        profiles=parquet_pycytominer_normalized,
```
you currently are pointing to the path, but would it be faster to point to the actual data you're processing above? (oh, but maybe that variable isn't guaranteed to be populated b/c of the `if not` statement)

d33bs replied:
Thanks for checking on this. During implementation I found that when I used output settings to export to Parquet that `None` was returned (we don't receive the DataFrame) from pcytominer. There were also kernel crashes during early troubleshooting with datasets that couldn't fit into the system I was using which made for long time periods of reprocessing. As a result I structured this to always read from file for intermediate steps.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+        df_input (pd.DataFrame]):
+            A dataframe which is expected to contain
+            numeric columns to be used for UMAP fitting.
+        cols_metadata (List[str]):
```
```suggestion
        cols_metadata_to_exclude (List[str]):
```

This adds a bit more clarity

d33bs replied:
Thanks, will add this in!

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    """
+
+    # Set constants
+    umap_n_components = 2
```
prob should be a default arg?

jenna-tomkinson replied:
Agreed. Also maybe have a random state set here too unless I missed it.

d33bs replied:
Ah good point! Thanks, will add in.

d33bs replied:
I initially left random state out of the UMAP configuration to help with speed of processing (setting random state disallows parallel processing). I've added parameter with a default of "None" for this to the function for now, but wanted to double check: should we stick with setting a random state here? @jenna-tomkinson - thoughts?

d33bs replied:
Will stick with a static random state here.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+                for col in df_input.columns.tolist()
+                if col not in metadata_cols and "cqc." not in col
+            ]
+        ].select_dtypes(include=[np.number])
```
hmm, i'm not sure what this does. Important to include?

d33bs replied:
Thanks for this comment - it selects numeric columns to avoid errors with calculations. This was an additional safety added to avoid using incompatible columns. I'll add a comment here to help clarify.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    df_input=pd.read_parquet(parquet_pycytominer_feature_selected),
+    cols_metadata=metadata_cols,
+)
+embeddings_with_outliers
```
rather than print everything, I recommend this:

```python
print(embeddings_with_outliers.shape)
embeddings_with_outliers.head(3)
```

It makes things a bit more manageable to render and digest

d33bs replied:
Thanks, will add this in.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+
+
+# +
+def plot_hvplot_scatter(
```
pep

d33bs replied:
Thanks, I'll move this to earlier in the notebook.

gwaybio reviewed (commented):
Love this analysis! I made some (mostly minor) comments to think about and potentially improve the work

jenna-tomkinson commented on docs/src/examples/images/umap_large_nuclei_outliers_BR00117012.png:
Super super tiny!

d33bs replied:
Thanks, will re-render this to help shrink the width (which is what makes this a appear small in GitHub).

jenna-tomkinson commented on docs/src/examples/images/umap_small_and_low_formfactor_nuclei_outliers_BR00117012.png:
Same comment on being tiny lol

d33bs replied:
Thanks, will re-render this to help shrink the width (which is what makes this a appear small in GitHub).

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+
+# # JUMP UMAP analysis with coSMicQC
+#
+# This notebook analyzes [JUMP](https://jump-cellpainting.broadinstitute.org/) data by leveraging [UMAP](https://arxiv.org/abs/1802.03426) and [coSMicQC](https://github.com/WayScience/coSMicQC).
```
Recommend including that this is JUMP cpg0000 pilot data since I don't see it explicitly stated below.

d33bs replied:
Thanks, we link to the exact file below, and I'll add in a note about which JUMP dataset on this line.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+# ## Outline
+#
+# We focus on a single file from the JUMP dataset: [`BR00117012.sqlite`](https://open.quiltdata.com/b/cellpainting-gallery/tree/cpg0000-jump-pilot/source_4/workspace/backend/2020_11_04_CPJUMP1/BR00117012/BR00117012.sqlite).
+# This file is downloaded and prepared by [CytoTable](https://github.com/cytomining/CytoTable) to form a single-cell [Parquet](https://parquet.apache.org/) file which includes all compartment feature data.
```
```suggestion
# This file is downloaded and prepared by [CytoTable](https://github.com/cytomining/CytoTable) to form a single-cell [Parquet](https://parquet.apache.org/) file which includes all compartment feature data at the single-cell level.
```

d33bs replied:
Thanks, will add this in.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+# Afterwards, we use UMAP to demonstrate patterns within the data.
+#
+
+# + editable=true slideshow={"slide_type": ""}
```
Is this an artifact of something or is this needed?

d33bs replied:
Thanks, this appears due to how the notebook is automatically exported using [jupytext](https://jupytext.readthedocs.io/en/latest/) (which saves time and reduces human-error from manually exporting the files separately). I'll see if there's a way to address this, but generally what's there is due to how the data are exported from JSON form.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+# avoid displaying plot export warnings
+logging.getLogger("bokeh.io.export").setLevel(logging.ERROR)
+
+# +
```
Recommend removing this code comment with a `+` if it is not needed.

d33bs replied:
Thanks for checking on this - this is from the automatic export of the file, [see here for more information](https://github.com/WayScience/coSMicQC/pull/88#discussion_r1752656428).

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+        source_datatype="sqlite",
+        chunk_size=8000,
+        preset="cellprofiler_sqlite_cpg0016_jump",
+        no_sign_request=True,
```
Recommend adding code comment on with no_sign_request is.

d33bs replied:
Thanks - I'll add in a comment about this.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+        dest_datatype="parquet",
+        source_datatype="sqlite",
+        chunk_size=8000,
+        preset="cellprofiler_sqlite_cpg0016_jump",
```
Is there a reason that the preset uses `cpg0016`? I believe we have only ever processed data from cpg0000 so just wondering what went into this presets name.

d33bs replied:
Thanks - we use a test in CytoTable which leverages a small-ish file from `cpg0016` to validate operations. The preset is based on this dataset, which is why it shows that name. The patterns from `cpg0016` from a data standpoint appeared to persist enough that this worked just as well for `cpg0000`.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+# read only the metadata from parquet file
+parquet.ParquetFile(merged_single_cells).metadata
+
+# + editable=true slideshow={"slide_type": ""}
```
Okay what is this for? I have seen this before and made a comment on it so it must have a purpose lol

d33bs replied:
Thanks for checking on this - this is from the automatic export of the file, [see here for more information](https://github.com/WayScience/coSMicQC/pull/88#discussion_r1752656428).

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    "Image_TableNumber",
+]
+
+# read only metadata columns with outlier-specific columns
```
```suggestion
# read only metadata columns with feature columns used for outlier detection
```

d33bs replied:
Thanks, I'll add this in.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+].head()
+
+# +
+# create a column which indicates at least one outlier being found
+df_labeled_outliers["analysis.included_at_least_one_outlier"] = df_labeled_outliers[
+    [col for col in df_labeled_outliers.columns.tolist() if ".is_outlier" in col]
+].any(axis=1)
```
What does it mean for "at least one outlier being found"? Is this per condition? If the data is at a single-cell level, then wouldn't this just be "is_outlier"? I would recommend adding more background for this column.

d33bs replied:
Thanks for checking on this - this creates a column which indicates whether "any" outlier was found from all outlier threshold sets within coSMicQC. I did this to help summarize the results in a single plot, instead of showing for example a single threshold set. I'll add in a few more notes to this effect.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,421 @@
+    "analysis.included_at_least_one_outlier"
+].value_counts()
+
+# prepare data for normalization and feature selection
+# by removing cosmicqc and analaysis focused columns.
+df_for_normalize_and_feature_select = df_feature_selected_with_cqc_outlier_data[
+    [
+        col
+        for col in df_feature_selected_with_cqc_outlier_data.columns.tolist()
+        if "cqc." not in col and "analysis." not in col
+    ]
+]
```
Do you end up adding back in these columns? Or how are you able to label by outlier status?

d33bs replied:
Thanks for checking on this - I used the indexes from the data to help relate this back to the data points later. The scatter plots use the index location as a way to relate this to the data.

jenna-tomkinson reviewed (approved):
I am approving this as I don't see this needing much more changes.

I am a bit confused by how the embeddings are connected back to the cosmicqc outlier status. Maybe more clarity would help me understand, since to me that could be impacting the pattern in the UMAPs. 

Other than that, exciting to get this created! Feel free to merge or ask for further feedback after addressing comments.

d33bs commented on CITATION.cff:
```diff
@@ -95,3 +95,68 @@ references:
+        given-names: Anne E.
+    date-accessed: "2024-08-21"
+    title: >-
+      Three million images and morphological profiles of cells treated with matched chemical and genetic perturbations
```
This comes directly from the publication citation mentioned in https://github.com/broadinstitute/cellpainting-gallery?tab=readme-ov-file#available-datasets for cpg0000-jump-pilot (https://doi.org/10.1038/s41592-024-02241-6). Should I use something else instead?

jenna-tomkinson replied:
Nope all good!

d33bs commented on docs/src/examples/images/umap_elongated_nuclei_outliers_BR00117012.png:
Thanks for catching this - these seem to have been exported with an extremely large width - possibly due to the default on the system used (which leverages two monitors). I'll see about making this a bit smaller. Generally, as a sidenote, the drivers for exporting are based on either Chromium or Firefox, and render through Selenium - making things a bit more complicated but not unresolvable.

d33bs replied:
Addressed by adding argument here in reference from this issue's workaround notes: https://github.com/holoviz/holoviews/issues/4489

d33bs commented on docs/src/examples/images/umap_BR00117012.png:
Thanks, I'll see about adding that level of detail in. The legend here represents density of points within the scatter plot.

d33bs replied:
I tried looking for a simple solution to this and couldn't find one. As a result, I've created https://github.com/WayScience/coSMicQC/issues/90 to help keep track of this comment in future revisions. Aside, this dealt with titling the colormap rather than a legend, which is treated distinctly from legends / keys in `hvplot` and likely other systems which tie in (`bokeh`, `plotly` etc.).

d33bs commented on docs/src/examples/images/umap_erroneous_outliers_BR00117012.png:
Thanks, yes, that's accurate -  I'll adjust this to read "all erroneous outliers" instead of help clarify things.

d33bs commented:
Thanks @gwaybio and @jenna-tomkinson for your earlier reviews! I've added a bunch of changes based on what you mentioned and this could use another look when there's a moment.

gwaybio commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.py:
```diff
@@ -0,0 +1,665 @@
+    ),
+    color_column="combined_embedding_color_label",
+    bgcolor="white",
+    cmap="RdYlGn",
```
a couple comments here, that need not delay merging (you probably don't need to address either, since my understanding is the UMAP you will show in the poster is the `coolwarm` palette. 

- We should avoid Red/Green color palette to accomodate folks with this colorblindness
- There are only three categories correct? Then shouldn't we use a different legend type?
- In general, all the legends should be labelled.

d33bs replied:
Thanks! I'll correct the color usage with vision accessibility in mind.

> There are only three categories correct? Then shouldn't we use a different legend type?

Because of how these plots display points it's important to provide gradients to show "less" or "more" based on how much information a pixel can contain. That said, I think you're correct, that these should only display two colors with their matching gradients.

> In general, all the legends should be labelled.

Thanks for the added attention to this, I had trouble finding information on this (as per #90 ) but found that this is called something different in Bokeh plots today (colomap labels). I've added labels to the function and every scatter plot in a later commit which will be added.

gwaybio reviewed (approved):
LGTM! 🤩 

No need to address the following comment (and the three comments below), I think you're good to go:

FWIW, my "killer app" idea involved _aggregating_ data with and without outliers, then applying UMAP to both of these aggregated datasets.

jenna-tomkinson commented on docs/src/examples/images/umap_comparison_with_and_without_erroneous_outliers_BR00117012.png:
This plot is a lot from my perspective. What colors represent outliers versus not outliers? Recommend more information or simplifying.

d33bs replied:
Thanks, will add more here with a colormap label and description here. Generally I had a hard time telling the difference between some of the shapes in the UMAP embeddings with vs without outliers. This plot is an attempt to differentiate the forms by using color and overlapping them together. I recognize it's nonstandard and may not be completely understandable but wanted to make an attempt towards nuanced distinction inside the same plot, similar to how the "onion" view of image comparisons work in GitHub PR reviews.

jenna-tomkinson commented on docs/src/examples/images/umap_large_nuclei_outliers_BR00117012.png:
I like this pattern with the large nuclei!

jenna-tomkinson commented on docs/src/examples/images/umap_with_all_outliers_BR00117012.png:
Is the legend necessary since it looks like all dots are red?

d33bs replied:
Thanks for checking on this - the scale I think could still be important based on how hvplot works with scatter plots. The density of the points should be colored darker to help with visual cues on how dense an area can be. This avoids data challenges mentioned here: https://datashader.org/user_guide/Plotting_Pitfalls.html . While there might not be relevant coloring displayed, I feel we should still be honest about the settings used with a scale just in case.

jenna-tomkinson commented on docs/src/examples/jump_umap_analysis_with_cosmicqc.ipynb:
I am not able to review this on GitHub due to the size. Is there a way you can simplify this notebook to reduce the lines of code output that are causing it to be so large? Maybe this won't be a problem, but do we know if there will be any issues rendering on the docs?

d33bs replied:
Great point - I had to work at keeping this under the 100MB limit for non-LFS files. The notebook is large mostly due to the embedded hvplots which include many points. The benefit of this approach is that the plots will be dynamically rendered in GitHub Pages through the docs, making the example more interactive. The downside is that the file is large, which is the reason for moving to some images over the plots themselves in the notebook. The GitHub web interface isn't equipped to handle the display of these (we can hope for better rendering in the future...) but I tested a docs build locally and feel good about how this will render once up on GitHub Pages.

jenna-tomkinson commented on pyproject.toml:
```diff
@@ -115,7 +128,7 @@ omit = [
 
 # set dynamic versioning capabilities for project
 [tool.poetry-dynamic-versioning]
-enable = true
+enable = false
```
Since this is now set to false, does the comment need to change to reflect this?

d33bs replied:
Thanks for catching this! I'm not exactly sure how this got changed, yes it should be changed back.

jenna-tomkinson reviewed (approved):
LGTM! I left some minor comments for you to address but feel free to merge when ready!

d33bs commented:
Thanks so much @jenna-tomkinson and @gwaybio for your reviews! I've addressed all comments and added new figures based on the results. I'll now merge this in.
